### DISCUSSION TITLE
Custom components

### DISCUSSION THREAD

**Discussion by sswatson** [marimo-team/marimo#456] [Q&A] [answered]
I'm curious whether marimo provides a mechanism by which users can add their own UI components similar to the ones provided in `marimo.ui`.

**Comment by mscolnick** [comment, marked as answer]
@sswatson, we currently don't allow for custom components, but it is on our roadmap. We are waiting for us to nail down the plugin/component API so we don't break 3rd party plugins. We also want to try to encourage really good core components. 

Is there a particular component you are looking for? If its core, it is something we can look into adding to `marimo.ui`.

**Reply by sswatson** [reply]
Thanks for the reply! I suspect my use case is out of scope for the project, but I'll mention it anyway.

Basically, what I'd like to achieve is a Python-based replacement for Excel. Marimo ticks almost all of the boxes for this, but the one thing you can do easily in Excel and not in any notebook I'm aware of is create and edit tables. The Marimo table component lets you select rows, but if I understand correctly, it can't really be used like a spreadsheet for creating new data frames.

I see that the suggested workflow for this use case is to connect to Google Sheets, but (1) I don't love Google Sheets — that's why I'm looking for something else! — and (2) I find that the friction associated with switching between applications to be sufficiently large that I would usually end up just working with the spreadsheet instead. Plus, you lose reactivity across the application boundary (I assume), and of course reactivity is a big part of the appeal.

Actually, what I'm looking for might not be a component but rather another editing mode — one for data frames — to complement the existing Python and Markdown modes. That way the data frame contents would be persistent, which is what you would want in almost all cases.

With this feature, a user could comfortably replace Excel with a single application that is more powerful (thanks to the Python ecosystem), built on sounder abstractions (namely, data frames rather than arbitrary grids of cells), and with a more narrative-oriented format (sequential exposition/code rather than separate sheets).

Anyway, this is a really cool project. I look forward to seeing it continue to grow!

**Reply by mscolnick** [reply]
Hey @sswatson thanks for the feedback! And I really appreciate your explanation, I totally understand. I think both of these would be useful generally, and I think they could belong in `marimo` long term. The former (a component) fits easier in the model today whereas the latter (a different cell view) may be a bit more effort but seems powerful. I can talk about both case below:

### Spreadsheets/Editable DF as a component

I do think this would be generally valuable and would make a lot of sense to include something like `mo.ui.editable_dataframe` or `mo.ui.spreadsheet` where the `.value` could be the resulting dataframe. We can also find a way to make the edits persist.

### A dataframe cell

You are right that dataframes are a great and widely used abstraction. I can imagine something where you can import a local csv (or other external datasources), create a dataframe cell (likely Python under the hood, kinda how handle markdown) which allows edit and write back to the datasource.

----

I don't know how fast we will get to this within our own internal priorities. I can try to get you a better answer/estimate, but we plan to publish our roadmap publicly which would help clear this up as well.

Given this started with you looking to add your own components, is this something you would like to contribute to? I am happy onboard you to the repo, get you set up to run with this, and pair with you along the way. If that would be of interest, [join our Discord channel ](https://discord.gg/rAhpfKwuaN) and DM me, we can chat more there.

**Reply by sswatson** [reply]
FFR: I joined the Discord channel to continue this discussion.
